flow/layers/backdrop_filter_layer.cc

```diff
@@ -11,13 +11,57 @@ BackdropFilterLayer::BackdropFilterLayer(sk_sp<SkImageFilter> filter)
 
 BackdropFilterLayer::~BackdropFilterLayer() = default;
 
+void BackdropFilterLayer::initRetained(std::shared_ptr<Layer> retainedLayer) {
+  BackdropFilterLayer* bdfLayer = (BackdropFilterLayer*)retainedLayer.get();
+  if (bdfLayer->retainedBackdrop_) {
+    retainedBackdrop_ = bdfLayer->retainedBackdrop_;
+  } else {
+    retainedBackdrop_ = std::move(retainedLayer);
+  }
+}
+
+void BackdropFilterLayer::Preroll(PrerollContext* context,
+                                  const SkMatrix& matrix) {
+  ContainerLayer::Preroll(context, matrix);
+
+  if (retainedBackdrop_ && context->raster_cache &&
+      context->view_embedder == nullptr) {
+    context->raster_cache->PrepareForSnapshot(context, retainedBackdrop_.get());
+  }
+}
+
 void BackdropFilterLayer::Paint(PaintContext& context) const {
   TRACE_EVENT0("flutter", "BackdropFilterLayer::Paint");
   FML_DCHECK(needs_painting());
 
-  Layer::AutoSaveLayer save = Layer::AutoSaveLayer::Create(
-      context,
-      SkCanvas::SaveLayerRec{&paint_bounds(), nullptr, filter_.get(), 0});
+  bool need_backdrop_paint = true;
+  bool need_snapshot = false;
+  if (retainedBackdrop_ && context.surface &&
+      context.view_embedder == nullptr) {
+    RasterCacheResult my_cache =
+        context.raster_cache->Get(retainedBackdrop_.get(), SkMatrix::I());
+    if (my_cache.is_valid()) {
+      my_cache.drawSnapshot(*context.leaf_nodes_canvas);
+      need_backdrop_paint = false;
+    } else {
+      need_snapshot = true;
+    }
+  }
+
+  if (need_backdrop_paint) {
+    Layer::AutoSaveLayer save = Layer::AutoSaveLayer::Create(
+        context,
+        SkCanvas::SaveLayerRec{&paint_bounds(), nullptr, filter_.get(), 0});
+  }
```

Comment:
This is changing how it draws.

Previously, it
1. calls `saveLayer` with a backdrop filter (by creating `AutoSaveLayer`)
2. paint children to the offscreen layer
3. calls `restore` to filter the backdrop, and composite the offscreen layer on top of it (by destructing `AutoSaveLayer`).

Now, it
1. calls `saveLayer` with a backdrop filter (by creating `AutoSaveLayer`)
2. calls `restore` to filter the backdrop, and composite the empty offscreen layer on top of it (by destructing `AutoSaveLayer`).
3. paint children to the on-screen surface.

Below are 2 fiddles that illustrate “offscreen paint with saveLayer” and “on-screen paint without saveLayer” could have different semantics:
- https://fiddle.skia.org/c/83ed46ceadaf90f36a4df3b98cbe1c35
- https://fiddle.skia.org/c/704acfa049a7e99fbe685232c45d1582

Reply:
Which is correct, though. The BDF widget says that it applies its filter to the existing background and ***then*** paints the child. The new code is following that description more accurately. It forces the filter to complete before the child is painted.

The description doesn't really indicate that it paints the child on top of the filtered output and then composes the filtered output into the scene together.

On a practical note, I can't snapshot the results until I take the ASL out of scope, so I can't grab the output of the blur until I've done the restore. Since a common use case may be a non-static child on top of a blurred static background, having to cache the entire rendering including the child will mean they either can't use the flag if there is anything that will change in the child, or we would have to recache if the child changes, which could be fairly frequent since the child is probably the only thing "happening" in the app at that moment (otherwise, why blur out everything below it?).

Reply:
In case that the new implementation in the correct thing to do, we need to file a breaking change notification.

Also, if we only want to blur the background without creating a new layer, a saveLayer seems to be an over-kill as we're creating and compositing an unnecessary new and empty layer. I'm not sure if Skia has already optimized this case. CC @bsalomon . I'm also going to ask in the Skia-Flutter chat room.

I understand your "ASL out of scope" concern. One way to deal with a dynamic children and not changing our current semantics is:
1. Create ASL with backdrop filter
2. Let ASL get out of scope, snapshot the background
3. Create a new saveLayer command without backdrop filter
4. paint children
5. restore

In that case, the saveLayer with backdrop is also an over-kill. So we'll have to figure that out anyway.